### ISSUE TITLE
[Experiment] Build Blackwell train-inference mismatch baseline for VIME post-training

### ISSUE THREAD

**Issue by inaniloquentee** [vllm-project/vime#365]
## Background

VIME already tracks train/rollout logprob consistency through `train/train_rollout_logprob_abs_diff`, which measures the mean absolute difference between Megatron recomputed logprobs and rollout-engine logprobs on the same sampled tokens.

Relevant prior context:

- #64 analyzed `train_rollout_logprob_abs_diff` between vime/vLLM and slime/SGLang.
- #107 includes GB200 post-training experiment notes for Qwen3-30B-A3B in colocate and non-colocate setups.

This issue tracks a new Blackwell/B-series experiment. The first step is to intentionally construct a measurable train-inference mismatch scenario on VIME and use it as the baseline. Later we will plug in Blackwell-specific operators/kernels and compare against the same baseline configuration.

## Goal

Create a reproducible B-card post-training baseline where the rollout path and training path have a controlled, measurable mismatch.

The baseline should answer:

1. Can VIME post-training run stably on B200/GB200 under this setup?
2. Can we observe a non-trivial `train/train_rollout_logprob_abs_diff` signal?
3. Is the mismatch stable enough to serve as an A/B baseline for future Blackwell operator integration?
4. After integrating B-card operators, can we compare correctness and performance under the same workload?

## Baseline Design

Use Megatron as the training backend and vLLM as the rollout backend on Blackwell GPUs.

To intentionally create a train-inference mismatch, start with one or more controlled differences between the training and rollout paths:

- BF16 Megatron training plus FP8 vLLM rollout/inference.
- Rollout-side logprobs from vLLM compared against Megatron recomputed logprobs.
- Sampling filters enabled on the rollout side, for example `top_p < 1.0` or `top_k > 0`.
- Optional follow-up: compare `processed_logprobs` against other vLLM logprob modes if available.

Recommended first baseline:

```text
Hardware: B200 / GB200
Model: Qwen3-30B-A3B
Training backend: Megatron
Rollout backend: vLLM
Training precision: BF16
Rollout/inference precision: FP8 if available
Algorithm: GRPO
Primary metric: train/train_rollout_logprob_abs_diff
```

This should produce a measurable mismatch while still being close to a realistic production post-training setup.

## Hardware Matrix

| Stage | Hardware | Mode | Purpose |
|---|---|---|---|
| P0 | 1 B200/GB200 node | colocate | Smoke test and config validation |
| P1 | 2 GB200 nodes, 8 GPUs total | colocate | Stable baseline |
| P2 | 4 GB200 nodes, 16 GPUs total | non-colocate | Production-like train/rollout split |
| P3 | same as P1/P2 | candidate B-card operator path | A/B comparison |

## Suggested Baseline Config

Model and data:

```text
--hf-checkpoint /path/to/Qwen3-30B-A3B-FP8
--ref-load /path/to/Qwen3-30B-A3B_torch_dist
--prompt-data /path/to/dapo-math-17k
--apply-chat-template
--input-key prompt
--label-key label
--rm-type deepscaler
```

Training and rollout:

```text
--train-backend megatron
--advantage-estimator grpo
--num-rollout 100
--rollout-batch-size 32
--n-samples-per-prompt 8
--global-batch-size 256
--rollout-max-response-len 8192
--rollout-temperature 1.0
--rollout-top-p 0.95
--rollout-top-k 40
--balance-data
```

Megatron parallelism:

```text
--tensor-model-parallel-size 4
--sequence-parallel
--pipeline-model-parallel-size 1
--context-parallel-size 2
--expert-model-parallel-size 8
--expert-tensor-parallel-size 1
--use-dynamic-batch-size
--max-tokens-per-gpu 8192
--recompute-granularity full
--recompute-method uniform
--recompute-num-layers 1
--attention-dropout 0.0
--hidden-dropout 0.0
--attention-softmax-in-fp32
--attention-backend flash
```

vLLM rollout:

```text
--rollout-num-gpus 8
--rollout-num-gpus-per-engine 2
--vllm-gpu-memory-utilization 0.8
--vllm-router-policy consistent_hash
--vllm-enable-expert-parallel
--vllm-all2all-backend deepep_high_throughput
```

For colocate mode:

```text
--colocate
--vllm-gpu-memory-utilization 0.5
```

## Metrics To Track

Primary correctness and mismatch metrics:

- `train/train_rollout_logprob_abs_diff`
- `rollout/log_probs`
- `rollout/rollout_log_probs`
- `train/ppo_kl`
- `train/pg_clipfrac`
- `train/loss`
- `rollout/raw_reward`

Performance metrics:

- total step time
- rollout generation time
- Megatron logprob recomputation time
- weight sync time
- training time
- tokens/sec
- peak GPU memory
- vLLM KV cache utilization, if available

Suggested summary table:

| Run | Hardware | Mode | Precision | Operator Path | Final Reward | logprob_abs_diff Mean | Step Time Mean | Notes |
|---|---|---|---|---|---|---|---|---|
| baseline | GB200 | colocate | BF16 train + FP8 rollout | default | TBD | TBD | TBD | mismatch baseline |
| candidate | GB200 | colocate | BF16 train + FP8 rollout | B-card operator | TBD | TBD | TBD | compare against baseline |
| baseline | GB200 | non-colocate | BF16 train + FP8 rollout | default | TBD | TBD | TBD | production-like split |
| candidate | GB200 | non-colocate | BF16 train + FP8 rollout | B-card operator | TBD | TBD | TBD | compare against baseline |

## Acceptance Criteria

The baseline is valid if:

- Training completes at least 30 rollout steps without NCCL, HTTP, CUDA illegal access, or OOM failures.
- `rollout/raw_reward` does not collapse.
- `train/train_rollout_logprob_abs_diff` is present in logs.
- The mismatch is non-trivial and measurable, ideally consistently above the near-aligned range observed in prior runs.
- Metrics are stable enough to compare against a later candidate run.
- Commit hash, Docker image, hardware type, run mode, exact config, and wandb/log links are recorded.

## Follow-up Candidate Work

After the baseline is established, integrate Blackwell/B-card specific operators and rerun the same config.

The candidate comparison should check:

1. Does the B-card operator reduce `train_rollout_logprob_abs_diff`?
2. Does it preserve or improve `rollout/raw_reward`?
3. Does it improve rollout throughput or logprob recomputation time?
4. Does it introduce numerical drift under BF16/FP8 mixed paths?
5. Does it behave consistently between colocate and non-colocate modes?

## Notes and Risks

- If mismatch is too small, increase the intentional rollout/training difference by enabling stronger sampling filters or using FP8 rollout.
- If mismatch is too large and reward collapses, reduce sampling filters before changing model parallelism.
- `top_p < 1` or `top_k > 0` can make rollout logprobs differ from Megatron replay logprobs because rollout-side processed logprobs may include sampling filter semantics.
- For GB200 multi-node runs, record NCCL, MNNVL, CUDA, PyTorch, vLLM, and Megatron environment details.
- Keep the first baseline simple. The immediate goal is not best reward, but a clean and measurable mismatch baseline for future B-card operator A/B comparison.
